### PR TITLE
[FLINK-11974][streaming] Introduce StreamOperatorSubstitutor to help table perform the whole Operator CodeGen

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

If we need CodeGen an entire Operator, one possible solution is to introduce an OperatorWrapper, then generate a CodeGen sub-Operator in OperatorWrapper's open, and then proxy all methods to the sub-Operator.
Doing so results in multiple virtual function calls, so we introduce a StreamOperatorSubstitutor

## Verifying this change

ut

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by JingsongLi** [comment]
cc @pnowojski  @sunhaibotb @KurtYoung

**Comment by KurtYoung** [comment]
@StephanEwen you may want to look at this. For table API & SQL, all codegen operators will implements `StreamOperatorSubstitutor`, and we ship the operator codes with it. In runtime, we will compile the actual stream operator when `getActualStreamOperator` is invoked.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorSubstitutor.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * @param <OUT> output type of the actual stream operator
+ */
+public interface StreamOperatorSubstitutor<OUT> {
+
+	/**
+	 * Produces the actual stream operator.
+	 *
+	 * @param userCodeClassLoader the user code class loader to use.
+	 * @return the actual stream operator created on {@code StreamTask}.
```
Comment (KurtYoung) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorSubstitutor.java]: the actual stream operator which will be used by {@code StreamTask}.

**Comment by wuchong** [comment]
LGTM. The implementation touches little to the runtime, I think it's safe to get it in.

**Comment by JingsongLi** [comment]
Hi @pnowojski I responded to you in JIRA.

**Pull request closed without merging**